### ISSUE TITLE
`jnp.sinc`'s gradient has large errors

### ISSUE THREAD

**Issue by howsiyu** [jax-ml/jax#34139]
### Description

The current `jnp.sinc` which uses
```python
return _where(eq_zero, _sinc_maclaurin(0, pi_x),
                lax.div(lax.sin(safe_pi_x), safe_pi_x))
```
has large errors on its gradients when `x` and thus `safe_pi_x` is close to zero.
To see why, note that

$$
\frac{d}{dx}\frac{\sin(x)}{x} = \frac{1}{x}\left( \cos(x) - \frac{sin(x)}{x} \right)
$$

When $x \approx 0$, Both $\cos(x)$ and $\frac{sin(x)}{x}$ are slightly smaller than 1 and have error of magnitude $u$ = `jnp.finfo(x.dtype).epsneg`. So the error of $\frac{d}{dx}\frac{\sin(x)}{x} \approx -\frac{x}{3}$ has absolute error of magnitude $\frac{u}{x}$ and relative error of magnitude $\frac{3u}{x ^ 2}$ which can be very large when $x \approx 0$.

We can solve the issue by using `_sinc_maclaurin` for values near 0 instead of just 0 and using higher order taylor series for `_sinc_maclaurin`. The truncated Maclaurin series of

$$
\frac{d}{dx}\frac{\sin(x)}{x} = \sum_{i=1}^{n} \frac{(-1)^i}{(2i+1)(2i-1)!} x^{2i-1}
$$

has error term of magnitude $\frac{x^{2n+1}}{(2n+3)(2n+1)!}$ when $\lvert x \rvert$ is small. The floating point error caused by addition can be ignored when $\lvert x \rvert$ is small and we sum from the right as each term is significantly larger than all the terms at its right. So we pick an integer $n$ and set the boundary to use the Maclaurin series to have $\frac{x^{2n+1}}{(2n+3)(2n+1)!} = \frac{u}{x}$ or equivalently $x = \left( (2n+3)(2n+1)! \ u \right) ^ {\frac{1}{2n+2}}$. Picking $n$ now is just a tradeoff between speed and accuracy. For most applications probably $n = 3$ is enough which gives $x = 0.47752283171543736$ and maximum relative error of 7.841751510599596e-07 for `float32` and $x = 0.038704406247251605$ and maximum relative error of 2.2233644923967434e-13 for `float64`.

This also solves https://github.com/jax-ml/jax/issues/10750 as `_sinc_maclaurin` always depends on $x$.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.8.2.dev20251209+f37438779
jaxlib: 0.8.1
numpy:  2.3.5
python: 3.13.0 (main, Oct 16 2024, 03:23:02) [Clang 18.1.8 ]
device info: NVIDIA L40S-4, 4 local devices"
process_count: 1
platform: uname_result(system='Linux', node='', release='6.8.0-86-generic', version='#87-Ubuntu SMP PREEMPT_DYNAMIC Mon Sep 22 18:03:36 UTC 2025', machine='x86_64')
```

**Comment by howsiyu** [comment]
The above analysis is only for precision of $\frac{d}{dx} \frac{sin(x)}{x}$ though obviously we care (more) about precision of $\frac{sin(x)}{x}$. We can use $n+1$ terms for the 0th derivative which would compute `sinc` up to machine precision in the selected range. For higher derivative the same phenomenon of cancelling and divide by $x$ occur again so the naive computation's precision is only going to get worse so we can use fewer terms to match the naive computation's precision at the boundaries.

**Comment by abhinav772007** [comment]
Hi,can I work on this issue?

**Comment by howsiyu** [comment]
I have already implemented the proposed update though I will like to hear the maintainer's opinion before submitting a pull request.

**Comment by jakevdp** [comment]
Assigning @mattjjj, who worked on the initial maclaurin series autodiff approach for `sinc`.

**Comment by yash-pal1** [comment]
Could you please assign this issue to me?

**Comment by vishalharkal15** [comment]
.take

**Comment by jonbarron** [comment]
I got it.
